File: src/scenario/ui/_exec.js
/*
 * Copyright 2020-2023 Alexis Royer <https://github.com/alxroyer/scenario>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


/**
 * @file
 * @brief Action executions.
 */


/** @var {object} `scenario.exec` package. */
scenario.exec = {};


// Execution.

// Install event listeners for execution buttons.
scenario.onLoad(() => {
    for (/** @var {HTMLElement} */ const _button of document.getElementsByClassName("exec button")) {
        _button.addEventListener("click", (e) => {
            e.preventDefault();

            scenario.exec._execute(_button.getAttribute("href"));
        });
    }
});

/**
 * @brief Executes a scenario action.
 * @param {string} url Action URL.
 * @returns {void}
 */
scenario.exec._execute = (url) => {
    // Process the given URL asynchronously.
    // Inspired from https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest_API/Synchronous_and_Asynchronous_Requests.
    console.debug(`Executing '${url}'...`);

    /** @var {XMLHttpRequest} */ const _req = new XMLHttpRequest();
    _req.open(
        "GET", url,
        true,  // Asynchronous request.
    );
    /** @var {string} */ let _title = `Request '${url}' error`;
    /** @var {string} */ let _message = "";
    _req.onload = (e) => {
        // Check state is DONE (see https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/readyState).
        if (_req.readyState !== 4) {
            _message = `Request '${url}' bad state ${_req.status}`;
            console.error(_message);
        } else {
            _message = `Request '${url}' returned ${_req.status} - ${_req.statusText}`;

            // Check HTTP status is OK.
            if (_req.status !== 200) {
                console.error(_message);
            } else {
                console.debug(_message);

                // Parse the JSON execution result.
                console.debug(`JSON content: '${_req.responseText}'`);
                /** @var {object} */ const _json = JSON.parse(_req.responseText);
                _title = _json.title;
                _message = _json.text;
            }
        }

        // Show execution result.
        scenario.exec.showResult(_title, _message);
    };
    _req.onerror = (e) => {
        scenario.exec.showResult(_title, _req.statusText);
    };
    _req.send();
};


// Execution result.

/** @var {boolean} Tells whether the `.exec-result` popup div shall be used. `alert()` called by default. */
scenario.exec.useExecResultPopupDiv = false;

/**
 * @brief Displays execution results.
 * @param {string} title Action title.
 * @param {string} text Execution result as plain text.
 * @returns {void}
 */
function scenarioShowExecResultPopup(title, text) {
    console.debug(`Displaying execution result popup with title='${title}' and text='${text}'`);

    /** @var {HTMLElement?} */ const _resultDiv = document.getElementById("exec-result");
    if (scenario.exec.useExecResultPopupDiv && _resultDiv) {
        // Set popup title.
        for (/** @var {HTMLElement} */ const _titleDiv of _resultDiv.getElementsByClassName("title")) {
            _titleDiv.textContent = title;
        }
        // Set popup content.
        for (/** @var {HTMLElement} */ const _textDiv of _resultDiv.getElementsByClassName("text")) {
            _textDiv.textContent = text;
        }

        // Display the popup.
        _resultDiv.style.display = "block";
    } else {
        // Display the message with `alert()`.
        alert(`[${title}]\n\n${text}`);

        // Refresh the page.
        scenario.exec._refreshCurrentPage();
    }
}

// Install the event listener for the OK button in the `.exec-result` popup div.
scenario.onLoad(() => {
    /** @var {HTMLElement?} */ const _resultDiv = document.getElementById("exec-result");
    if (_resultDiv) {
        for (/** @var {HTMLElement} */ const _button of _resultDiv.getElementsByClassName("exec-result button validate")) {
            _button.addEventListener("click", (e) => {
                e.preventDefault();

                // Hide the popup.
                console.debug("Hiding execution result popup");
                _resultDiv.style.display = "none";

                // Refresh the page.
                scenario.exec._refreshCurrentPage();
            });
        }
    }
});


// Refresh.

/**
 * @brief Refreshes the current page after execution result has been displayed.
 * @returns {void}
 */
scenario.exec._refreshCurrentPage = () => {
    console.debug(`Refreshing page '${window.location.href}'`);

    // Avoid reloading POST pages with parameters.
    // Inspired from https://stackoverflow.com/questions/1226714/how-to-get-the-browser-to-navigate-to-url-in-javascript#1226718
    //window.location.reload();
    window.location.href = window.location.href;
};

File: src/scenario/ui/_exec.py
# -*- coding: utf-8 -*-

# Copyright 2020-2023 Alexis Royer <https://github.com/alxroyer/scenario>
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
User interface action execution handler.
"""
import typing

import scenario

if True:
    from ._httprequesthandler import HttpRequestHandler as _HttpRequestHandlerImpl  # @inheritance
if typing.TYPE_CHECKING:
    from ._htmldoc import HtmlDocument as _HtmlDocumentType
    from ._httprequest import HttpRequest as _HttpRequestType


class Exec(_HttpRequestHandlerImpl):
    """
    Action execution request handler.

    Executes actions and returns a ``{"return_code": int, "title": str, "text": str}`` JSON result.
    """

    #: Base URL for action executions.
    _URL = "/execute"  # type: str

    class Arg(scenario.enum.StrEnum):
        """
        URL argument names.
        """
        #: Action argument.
        #:
        #: See :class:`Exec.Action` for possible values.
        ACTION = "action"

    class Action(scenario.enum.StrEnum):
        """
        :attr:`Exec.Arg.ACTION` values.
        """
        #: Reload the main requirement baseline.
        RELOAD_MAIN_REQ_BASELINE = "reload-main-req-baseline"
        #: Reload the campaign database.
        RELOAD_CAMPAIGN_DB = "reload-campaign-db"

    @staticmethod
    def mkurl(
            action,  # type: Exec.Action
            *,
            req_baseline=None,  # type: typing.Optional[scenario.ReqBaseline]
    ):  # type: (...) -> str
        """
        Builds an action execution URL.

        :param action: Action to create an URL for.
        :param req_baseline: Baseline to execute the action for. Main requirement baseline by default.
        :return: Action execution URL.
        """
        from ._reqbl import UI_REQ_BASELINES
        from ._httprequest import HttpRequest

        return HttpRequest.encodeurl(
            Exec._URL,
            args={
                **HttpRequest.mkurlargs(obj=req_baseline or UI_REQ_BASELINES.main),
                Exec.Arg.ACTION: action,
            },
        )

    @staticmethod
    def execresultdiv2html(
            html,  # type: _HtmlDocumentType
    ):  # type: (...) -> None
        """
        Generates the hidden ``.exec-result`` popup div.

        Hidden by default.
        Used by '_exec.js' to display execution results.

        :param html: HTML output page to feed.
        """
        from ._htmlgenbuttons import ButtonGenerator

        with html.addnode("div", id="exec-result", classes=["exec-result", "background"], style="display: none;"):
            with html.addnode("div", classes=["exec-result", "foreground"]):
                html.addnode("div", classes=["exec-result", "title"])
                html.addnode("div", classes=["exec-result", "text"])
                ButtonGenerator(html).addbutton(
                    classes=["exec-result", "validate"],
                    text="OK",
                )

    @staticmethod
    def actionbutton2html(
            html,  # type: _HtmlDocumentType
            request,  # type: _HttpRequestType
            action,  # type: Exec.Action
    ):  # type: (...) -> None
        """
        Creates an action execution button.

        :param html: HTML output page to feed.
        :param request: Current request being processed.
        :param action: Action to create a button for.
        """
        from ._htmlgenbuttons import ButtonGenerator
        from ._reqbl import UI_REQ_BASELINES

        _url = ""  # type: str
        _classes = ["exec"]  # type: typing.List[str]
        _text = "..."  # type: str
        _title = None  # type: typing.Optional[str]

        if action == Exec.Action.RELOAD_MAIN_REQ_BASELINE:
            # Don't display the `.reload-main-req-baseline` button if the page is not for it.
            if request.req_baseline is not UI_REQ_BASELINES.main:
                return

            _url = Exec.mkurl(Exec.Action.RELOAD_MAIN_REQ_BASELINE)
            _classes.append(Exec.Action.RELOAD_MAIN_REQ_BASELINE)
            _text = "Reload"
            _title = "Reload the main baseline (tests and requirements)"

        elif action == Exec.Action.RELOAD_CAMPAIGN_DB:
            _url = Exec.mkurl(Exec.Action.RELOAD_CAMPAIGN_DB)
            _classes.append(Exec.Action.RELOAD_CAMPAIGN_DB)
            _text = "Reload campaigns"
            _title = "Relaod the campaign database"

        else:
            raise ValueError(f"Unknown action {action!r}")

        ButtonGenerator(html).addbutton(
            href=_url,
            classes=_classes,
            title=_title,
            text=_text,
        )

    def __init__(self):  # type: (...) -> None
        """
        Configures the logger instance.
        """
        from ._debugclasses import UIDebugClass

        _HttpRequestHandlerImpl.__init__(self, UIDebugClass.EXEC)

    def process(
            self,
            request,  # type: _HttpRequestType
    ):  # type: (...) -> bool
        from .._scenarioconfig import SCENARIO_CONFIG
        from ._reqbl import UI_REQ_BASELINES

        # Filter `request`.
        if not request.base_path.startswith(Exec._URL):
            self.debug("Unexpected base path %r", request.base_path)
            self.debug("%r not processed", request)
            return False

        # Prepare return values.
        _title = ""  # type: str
        _return_code = scenario.ErrorCode.INPUT_MISSING_ERROR  # type: scenario.ErrorCode
        _text = f"'{Exec.Arg.ACTION}' parameter missing."  # type: str

        # Read ACTION parameter.
        _action = request.getarg(Exec.Arg.ACTION, default="")  # type: str

        # Execute actions.
        try:
            if _action == Exec.Action.RELOAD_MAIN_REQ_BASELINE:
                _title = "Reload main requirement baseline"

                UI_REQ_BASELINES.main = scenario.ReqBaseline.fromfiles(
                    name=UI_REQ_BASELINES.main.name,
                    req_db_paths=SCENARIO_CONFIG.reqdbpaths() or None,
                    test_suite_paths=SCENARIO_CONFIG.testsuitepaths() or None,
                    log_info=True,
                )

                _return_code = scenario.ErrorCode.SUCCESS
                _text = "Main requirement baseline reloaded successfully: "
                if UI_REQ_BASELINES.main.req_db.getallreqs():
                    _text += f"{len(UI_REQ_BASELINES.main.req_db.getallreqs())} requirements and "
                _text += f"{len(UI_REQ_BASELINES.main.scenarios)} tests loaded."

            elif _action == Exec.Action.RELOAD_CAMPAIGN_DB:
                _title = "Reload campaigns"

                scenario.campaign_db.load()

                _return_code = scenario.ErrorCode.SUCCESS
                _text = f"{len(scenario.campaign_db.campaign_executions)} campaigns reloaded successfully."

            else:
                _title = "Action error"
                _return_code = scenario.ErrorCode.INPUT_FORMAT_ERROR
                _text = f"Invalid '{Exec.Arg.ACTION}' parameter {_action!r}"

        except Exception as _err:
            _return_code = scenario.ErrorCode.fromexception(_err)
            _text = repr(_err)

        # Return result.
        _res = {
            "return_code": _return_code.value,
            "title": _title,
            "text": _text,
        }  # type: scenario.types.JsonDict
        self.debug("Execution result: %r", _res)
        request.sendjson(_res)
        return True
